packages/type-safe-api/src/construct/integrations/s3.ts

```diff
@@ -0,0 +1,179 @@
+  readonly role?: IRole;
+
+  /**
+   * The HTTP method to use when invoking the S3 bucket
+   * @default - integration method is used
+   */
+  readonly method?: Method;
+
+  /**
+   * The path override to use when invoking the S3 bucket
+   * @default - integration path is used
+   */
+  readonly path?: string;
+
+  /**
+   * The error integration response to use when the S3 bucket returns an error
+   * @default ErrorIntegrationResponses.catchAll()
+   */
+  readonly errorIntegrationResponse?: ErrorIntegrationResponse;
+}
+
+/**
+ * An S3 integration
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-s3.html
+ */
+export class S3Integration extends Integration {
+  private readonly bucket: IBucket;
+  private readonly role: IRole;
+  private readonly method?: Method;
+  private readonly path?: string;
+  private readonly errorIntegrationResponse?: ErrorIntegrationResponse;
+
+  constructor(props: S3IntegrationProps) {
+    super();
+
+    this.bucket = props.bucket;
+
+    if (props.role) {
+      this.role = props.role;
+    } else {
+      this.role = (this.bucket.node.tryFindChild("ExecutionRole") ??
+        new Role(this.bucket, "ExecutionRole", {
+          assumedBy: new CompositePrincipal(),
+        })) as IRole;
+    }
+
+    this.method = props.method;
+    this.path = props.path;
+    this.errorIntegrationResponse = props.errorIntegrationResponse;
+  }
+
+  /**
+   * Render the S3 integration as a snippet of OpenAPI
+   */
+  public render(props: IntegrationRenderProps): ApiGatewayIntegration {
+    return {
+      type: "AWS",
+      httpMethod: (this.method ?? props.method).toUpperCase(),
+      uri: bucketInvocationUri(this.bucket, this.path ?? props.path),
+      credentials: this.role.roleArn,
+      responses: {
+        default: {
+          statusCode: "200",
+          responseParameters: props.corsOptions
+            ? generateCorsResponseParameters(props.corsOptions)
+            : {},
+          responseTemplates: {},
+        },
+
+        ...(
+          this.errorIntegrationResponse ?? ErrorIntegrationResponses.catchAll()
+        ).render(props),
+      },
+    };
+  }
+
+  /**
+   * Grant API Gateway permissions to invoke the S3 bucket
+   */
+  public grant({ scope, api, method, path }: IntegrationGrantProps) {
+    this.role.grant(
+      new ServicePrincipal("apigateway.amazonaws.com", {
+        conditions: {
+          "ForAnyValue:StringEquals": {
+            "aws.SourceArn": Stack.of(scope).formatArn({
+              service: "execute-api",
+              resource: api.restApiId,
+              // Scope permissions to any stage and a specific method and path of the operation.
+              // Path parameters (eg {param} are replaced with wildcards)
+              resourceName: `*/${method.toUpperCase()}${path.replace(
+                /{[^\}]*\}/g,
+                "*"
+              )}`,
+            }),
+          },
+        },
+      })
+    );
```

Comment:
I'm not sure if this will actually do anything as it's a grant on the role arn with no actions :)

https://github.com/aws/aws-cdk/blob/7c497ee998976b69ad1e2cd3e68f30cb593f8ded/packages/aws-cdk-lib/aws-iam/lib/role.ts#L573-L580

I think you want `grantAssumeRole`? :)

Reply:
Absolutely!